Add Style Guide (#1713)
exercism/problem-specifications: STYLE-GUIDE.md

# Exercism's Style Guide

This document acts as a Style Guide for the language and wording used in exercises.

## General principle

> All content should be written in US english. 

Reviewer: ```suggestion
All content should be written in US English. 
```

Reviewer 2: Maybe:

```
Regrettably, all content should be written in Simple English, colloquially known as US English.
```

Just kidding of course. But we might want to consider linking to a document that has the most prominent differences between English and US English, as they are **not** apparent to everyone.

Reviewer: I remember there being discussions on Slack about the use of `pants` vs `trousers` where @iHiD veto'd the use of US English. Perhaps that should be added as an exception to this rule? 🙃

Author: I'm happy to have "no pants" as a documented rule!

Reviewer 3: Dogs have pants, people have trousers.

Author: https://github.com/exercism/problem-specifications/issues/1715

> All content should be written in US english. 

Reviewer: Can we add a rule that all units must be SI units or SI-compatible, though?

Reviewer 2: Yes: https://en.wikipedia.org/wiki/Non-SI_units_mentioned_in_the_SI + SI please.

Author: Yep. Sounds good. (This is actually a potential example where the "exception" thing mentioned below fits in - if an example is about converting from one unit to another and we deliberately choose not to use `si` units).

@SaschaMann Either add a suggestion to this conversation or follow up with a seperate PR (the later is probably preferable so we can get this one over the line quicker)

Reviewer: ```suggestion
All content should be written in US English.
All units of measurement must be [SI](https://en.wikipedia.org/wiki/International_System_of_Units) or [SI-derived](https://en.wikipedia.org/wiki/SI_derived_unit) units.
```

> In the future other translations may occur, but the "official" Exercism natural langauge is US English.

Reviewer: ```suggestion
In the future other translations may occur, but the "official" Exercism language is US English.
```

> ## Explain or substitude mathematical and esotoric terms.

Reviewer 4: ```suggestion
## Explain or substitute mathematical and esoteric terms.
```
Spelling....

> ## Explain or substitude mathmatical and esotoric terms.

Reviewer: ```suggestion
## Explain or substitude mathematical and esotoric terms.
```

> ## Explain or substitude mathmatical and esotoric terms.

Reviewer: Could you perhaps add an example of a non-maths esoteric term? The content of this section only mentions maths terms.

Author: I can't think of one off the top of my head. But I'll try and come up with one and PR it. Feel free to do the same if you can think of one.

Reviewer 3: I would guess that if I were talking about "lower case" it would be the beer that is lower than the other case of beer.  While saying "lowercase" is talking about letters.

In any place that mathematical terms are used they should be explained or substituted out for terms that require less domain knowledge.

Examples:

> - Rather than using "natural numbers", we should use "positive integers" or "zero and positive integers". 

Reviewer: `zero`, `positive` and `integers` are mathematical terms.

Reviewer 5: But they are more plainly understood than `natural numbers` which appears to be the intent

Reviewer: I'm not convinced that "integers" is more plainly understood than "natural numbers".

Reviewer 5: ¯\\\_(ツ)_/¯

Author: > I'm not convinced that "integers" is more plainly understood than "natural numbers".

I'm pretty convinced it is as `Integer` is a commonly used class / type in most languages, whereas `natural` isn't a programming "concept" I've ever used.

But I'd be happy to say "positive whole numbers" too.

Reviewer: Yea that's precisely the thing. I hear the term integers mostly in a programming context, while natural numbers are used everywhere and more likely to be known even by programming novices because it's a term they (probably) heard in school.

As you know, I'm really not a fan of singling out maths as some scary voodoo thing that people must be afraid of. I'd be happier if it was an example where, instead of a mathematical term, an everyday term would be used. Maybe something like `Rather than using "bijective map", we should use "1-to-1 map"`? (Not a great example either.)

> But I'd be happy to say "positive whole numbers" too.

That's worse imo because positive whole numbers are just natural numbers :D

Author: I'm happy to use ~zero~ no esoteric terms (where integer might be considered esoteric here). Especially as it's kinder to non-native speakers.

But, we do actively presume someone has a programming background, and we equally actively presume someone doesn't have a maths background. So we should be more comfortable using programming words than maths words.

> That's worse imo because positive whole numbers are just natural numbers :D

Sometimes natural numbers include zero, sometimes they don't :P

All concepts have words that label them. But requiring someone to know more words is a less kind thing than using two words to explain something when they might not.

Saying 
> positive whole numbers (know as "natural numbers")

Then using `natural numbers` throughout the rest of the description is fine.

Reviewer 2: Isn't it good for understanding to do "whole numbers" (integers)" and "whole positive numbers (unsigned integers)"? If we're trying to be as _accessible_ as possible, this might be preferable.

My problem with _integer_ is that it's NOT the same thing across programming languages.

Reviewer: > But, we do actively presume someone has a programming background, and we equally actively presume someone doesn't have a maths background. So we should be more comfortable using programming words than maths words.

I'd argue that high school words are an even lower requirement than programming words, therefore we should use those when possible.

That said, this is kinda pointless, maybe we can come up with a differenti example instead of arguing whether integers or natural numbers are a more commonly known term?

Reviewer: ```suggestion
- Rather than using "bijective map", we should use "1-to-1 map".
```

Author: I (genuinely) don't understand "1 to 1 map".

Wikipedia says:

> In mathematics, a bijection, bijective function, one-to-one correspondence, or invertible function, is a function between the elements of two sets, where each element of one set is paired with exactly one element of the other set, and each element of the other set is paired with exactly one element of the first set.

If we needed to use the word bijection, we would need to be using some level of information that covers this. Maybe "where each element in one set has a paired element in another set"? (I don't know if that's correct - but that's how I'm interpretting that ^^)

```suggestion
- Rather than using "natural numbers", we should use "positive whole numbers".
```

Reviewer 6: I think if a bijection were required there would be some easy contextual way of describing the property (maybe we need things to be reversible, or maybe we just need a unique name for each robot and for there to be no names that don't correspond to robots). So this could be covered under "no jargon".

In general, I think "a one-to-one mapping" or "one-to-one correspondence" is more clear. If someone starts talking about surjections and bijections then I almost always just have to look it up. And btw, your definition is slightly wrong in that it doesn't ensure that the second set only contains elements paired with the first, nor that each element in the first set is paired with a unique element in the second.

This table is a nice way of explaining it for me:

![image](https://user-images.githubusercontent.com/6000761/96198371-8ea45e80-0f4c-11eb-8d8e-d8043fb0e228.png)

> - If we want to use the phrase "rational numbers", it must be explained in the introduction to the exercise.

Reviewer: Is a link to an explanation sufficient?

Author: I think at the very least, I'd want an example of what is and is not a rational number, and any information that is needed to solve the exercise. For rational numbers, I'd probably have:

> This exercise is about Rational Numbers: Numbers that can be expressed as a fraction of two whole numbers. e.g. 1/4 or 12/6. You can learn more about rational numbers at [maths.com](...).

If I needed to know about signs, I'd add whether rational numbers can be negative or not. If I didn't need to know for the exercise, I wouldn't mention it.

Reviewer: Thanks!

Author: Feel free to PR that clarification to the guide, etc 🙂

> - Rather than using the word range (which can have different meanings in different contexts) use "greater than x and less than y".

Reviewer: Depending on the context `x < _ < y` may be better (as in more readable), imo.

Reviewer 2: I prefer both.

```
x < ? < y (greater than x and less than y)
```

Author: ```suggestion
- Rather than using the word range (which can have different meanings in different contexts) use "x < ? < y (greater than x and less than y)".
```

## Use consistency within an exercise.

> There are some terms that have multiple valid spellings (e.g. "lower case" vs "lowercase"). 

Reviewer: Perhaps this could be the first example of an agreed spelling?

Author: Sure. But I also want a rule that explains what happens when multiple valid spellings are used without a documented decision in this guide (ie, they should be exercise-consistent)

Reviewer 4: lowercase seems to be more popular:
![image](https://user-images.githubusercontent.com/6978003/96163316-c0370e80-0ee7-11eb-8c17-3c844566476e.png)

https://english.stackexchange.com/questions/59409/lowercase-lower-case-or-lower-case

Reviewer 2: Good enough for me 👍

Reviewer 7: More popular != correct 😜

Author: I'd agree with lowercase, but I think this falls into the "it doesn't matter as long as its consistent within an exercise" bucket. Unless anyone feels strongly about this particular word, and wants to encode `lowercase` as a rule and find me another example for this bit :)

# Exercism's Style Guide

This document acts as a Style Guide for the language and wording used in exercises.

## General principle

All content should be written in US English.
All units of measurement must be [SI](https://en.wikipedia.org/wiki/International_System_of_Units) or [SI-derived](https://en.wikipedia.org/wiki/SI_derived_unit) units.
In the future other translations may occur, but the "official" Exercism language is US English.

> ## Explain or substitute mathematical and esoteric terms.

Reviewer 7: Should there be a similar rule for abbreviations, acronyms and initialisms ?

For example, a rule that is often used in academic writing is that an initialism should have its expansion in brackets immediately following its first use and thereafter can be used freely.

Reviewer 6: I think we should have a no jargon rule. And don't use initialisms etc at all unless they are better known by the initialism (e.g. HTTP, HTML, radar) or if the text would be ridiculously verbose otherwise.

Reviewer 7: In favour of no jargon. 
However, outside of the specific language related initialism, should we assume that all users will have knowledge of what I or you might consider better known ones, even for non-native US English speakers?

Yet we would still need to use initialisms in some cases. If we, hypothetically, agree that RADAR is jargon to non-US English speakers, then we would not have to be forced to use the long form each and every time it is mentioned, especially if it might form the name of test cases or some other oft repeated part of an exercise.

thus we need some kind of middle solution (hence my suggestion 😉).

I can think of a host of these that might be used depending on the exercise and track: FIFO, NASA, ABS, GIF, DIY, SONAR, SUV, UFO, CRUD, ANSI, API, ASCII, ICANN, CMYK, HTTP, FTP, DPI, DSL, JPEG, ... these are not especially unusual in programming or jargon-y, yet I can easily say that not everyone around the world would be familiar with all them

Author: Let's have a seperate PR and standalone discussion for that. I've created an issue 🙂 

https://github.com/exercism/problem-specifications/issues/1716
